packages/magic/src/index.ts

```diff
@@ -0,0 +1,206 @@
+          chainId: parseInt(chains[0].id, 10),
+          rpcUrl: chains[0].rpcUrl
+        }
+
+        let magicInstance = new Magic(apiKey, {
+          network: customNodeOptions
+        })
+
+        const loginWithEmail = async (emailAddress: string) => {
+          try {
+            await magicInstance.auth.loginWithMagicLink({ email: emailAddress })
+            return await magicInstance.user.isLoggedIn()
+          } catch (err) {
+            throw new Error(
+              `$An error occured while connecting your Magic wallet, please try again: ${err}`
+            )
+          }
+        }
+
+        const handleLogin = async () => {
+          loggedIn = await loginModal({
+            walletName: walletName,
+            brandingHTMLString: brandingHTML,
+            emailLoginFunction: loginWithEmail
+          })
+        }
+
+        if (!loggedIn) await handleLogin()
+
+        let magicProvider = magicInstance.rpcProvider
+        let provider: EIP1193Provider
+
+        function patchProvider(): EIP1193Provider {
+          provider = createEIP1193Provider(magicProvider.enable(), {
+            eth_requestAccounts: async () => {
+              try {
+                if (!loggedIn) await handleLogin()
+                const accounts = await magicProvider.request({
+                  method: 'eth_accounts'
+                })
+                return accounts
+              } catch (error) {
+                console.error('error in request accounts', error)
+                const { code } = error as { code: number }
+                if (code === -32603) {
+                  throw new ProviderRpcError({
+                    code: ProviderRpcErrorCode.ACCOUNT_ACCESS_REJECTED,
+                    message: 'Account access rejected'
+                  })
+                }
+                return []
+              }
+            },
+            eth_selectAccounts: null,
+            eth_getBalance: async () => {
+              const address = (
+                await magicProvider.request({
+                  method: 'eth_accounts'
+                })
+              )[0]
+              const balance = await magicProvider.request({
+                method: 'eth_getBalance',
+                params: [address, 'latest']
+              })
+              return balance
+                ? BigNumber.from(balance).mul('1000000000000000000').toString()
+                : '0'
+            },
+            wallet_switchEthereumChain: async ({ params }) => {
+              const chain = chains.find(({ id }) => id === params[0].chainId)
+              if (!chain) throw new Error('Chain must be set before switching')
+              currentChain = chain.id
+              // re-instantiate instance with new network
+              customNodeOptions = {
+                chainId: parseInt(chains[0].id, 10),
+                rpcUrl: chains[0].rpcUrl
+              }
+
+              magicInstance = new Magic(apiKey, {
```

Comment:
Did you get a chance to check if we need to do any cleanup on the original instance and or provider to prevent any possible memory leaks?

Reply:
We should be good as is without causing any extra calls